cluster/src/main/java/org/apache/iotdb/cluster/log/manage/serializable/SyncLogDequeSerializer.java

```diff
@@ -0,0 +1,310 @@
+import java.nio.channels.FileChannel;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import org.apache.iotdb.cluster.config.ClusterDescriptor;
+import org.apache.iotdb.cluster.exception.UnknownLogTypeException;
+import org.apache.iotdb.cluster.log.Log;
+import org.apache.iotdb.cluster.log.LogParser;
+import org.apache.iotdb.db.conf.IoTDBDescriptor;
+import org.apache.iotdb.db.engine.fileSystem.SystemFileFactory;
+import org.apache.iotdb.tsfile.utils.ReadWriteIOUtils;
+import java.util.List;
+
+public class SyncLogDequeSerializer implements LogDequeSerializer {
+
+  File logFile;
+  File metaFile;
+  FileOutputStream logOutputStream;
+  FileOutputStream metaOutputStream;
+  LogParser parser = LogParser.getINSTANCE();
+  private Deque<Integer> logSizeDeque = new ArrayDeque<>();
+  private LogManagerMeta meta;
+  // mark first log position
+  private long firstLogPosition = 0;
+  // removed log size
+  private long removedLogSize = 0;
+  // when the removedLogSize larger than this, we actually delete logs
+  private long maxRemovedLogSize = ClusterDescriptor.getINSTANCE().getConfig()
+      .getMaxRemovedLogSize();
+
+  /**
+   * for log tools
+   * @param logPath log dir path
+   */
+  public SyncLogDequeSerializer(String logPath) {
+    logFile = SystemFileFactory.INSTANCE
+        .getFile(logPath + File.separator + "logData");
+    metaFile = SystemFileFactory.INSTANCE
+        .getFile(logPath + File.separator + "logMeta");
+    String name = logFile.getAbsolutePath();
+    init();
+  }
+
+  /**
+   * log in disk is
+   * size of log | log buffer
+   * meta in disk is
+   * firstLogPosition | size of log meta | log meta buffer
+   */
+  public SyncLogDequeSerializer() {
+    String systemDir = IoTDBDescriptor.getInstance().getConfig().getSystemDir();
+    logFile = SystemFileFactory.INSTANCE
+        .getFile(systemDir + File.separator + "raftLog" + File.separator + "logData");
+    metaFile = SystemFileFactory.INSTANCE
+        .getFile(systemDir + File.separator + "raftLog" + File.separator + "logMeta");
+    init();
+  }
+
+  // only for test
+  public void setMaxRemovedLogSize(long maxRemovedLogSize) {
+    this.maxRemovedLogSize = maxRemovedLogSize;
+  }
+
+  public Deque<Integer> getLogSizeDeque() {
+    return logSizeDeque;
+  }
+
+  // init output stream
+  private void init() {
+    try {
+      if (!logFile.getParentFile().exists()) {
+        logFile.getParentFile().mkdir();
+        logFile.createNewFile();
+        logFile.createNewFile();
+      }
+
+      logOutputStream = new FileOutputStream(logFile, true);
+      metaOutputStream = new FileOutputStream(metaFile, true);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  @Override
+  public void addLast(Log log, LogManagerMeta meta) {
+    ByteBuffer data = log.serialize();
+    int totalSize = 0;
+    // write into disk
+    try {
+      totalSize += ReadWriteIOUtils.write(data, logOutputStream);
```

Comment:
It is a little misleading to use the word "total" and the operator "+=", it seems that you are adding some things up, but actually there is only one item.

Reply:
fixed~